### DISCUSSION TITLE
How to use materialize_to_memory with dependencies?

### DISCUSSION THREAD

**Discussion by EtienneT** [dagster-io/dagster#14630] [Q&A] [answered]
I want to write tests to be able to debug my assets in vscode.

I have a weekly partitionned asset that depends on multiple daily partitionned assets.

I want to test the weekly partitionned asset.  materialize_to_memory seems to be the solution and from the documentation you should be able to pass SourceAssets, but when I do something like this I get errors:

materialize_to_memory([price_weekly, SourceAsset("price_daily"), SourceAsset("categories_daily")],  partition_key="2023-05-21")

`dagster._core.errors.DagsterInvariantViolationError: Attempting to access run_id, but it was not provided when constructing the OutputContext`

**Comment by sryza** [comment]
Would you be able to provide a minimal examples that reproduces this issue?

**Reply by EtienneT** [reply]
I have three assets, two daily partitionned, one not partitionned like so:

```python
from dagster import (
    DailyPartitionsDefinition,
    asset,
    TimeWindowPartitionsDefinition,
    TimeWindowPartitionMapping,
    WeeklyPartitionsDefinition
)

@asset(
    partitions_def=DailyPartitionsDefinition(start_date="2023-01-01", end_offset=-3)
)
def price_final_daily(context) -> pd.DataFrame:
    date_str = context.asset_partition_key_for_output()
    date = pd.to_datetime(date_str)
	
	// code that generate a dataframe
    
    return df

@asset(
    partitions_def=DailyPartitionsDefinition(start_date="2023-01-01")
)
def price_recent_daily(context) -> pd.DataFrame:
    date_str = context.asset_partition_key_for_output()
    date = pd.to_datetime(date_str)
	
	// code that generate a dataframe
    
    return df


@asset()
def price_today_daily(context) -> pd.DataFrame:
	// code that generate a dataframe
	return df
```

Then I also have a weekly partionned asset
	
```python
weekly = WeeklyPartitionsDefinition(start_date="2023-01-01", end_offset=-1)

@asset(
    partitions_def=weekly
)
def price_weekly(context, price_final_daily, price_recent_daily, price_today_daily) -> pd.DataFrame:
    df = pd.concat([price_final_daily, price_recent_daily, price_today_daily])
    df = df.sort_values(['Time'])
    return df
```

Finally I just wanted to test how weekly assets that depends on daily assets gets their data 
	
```python
def test_price_weekly():
    result = materialize_to_memory([SourceAsset("price_final_daily"), SourceAsset("price_recent_daily"), SourceAsset("price_today_daily"), price_weekly],  partition_key="2023-05-21")
    df = result.output_for_node("price_weekly")
```

Executing this test, this is where I get the error.

```
dagster._core.errors.DagsterInvariantViolationError: Attempting to access run_id, but it was not provided when constructing the OutputContext
```

I know that I could call price_weekly directly and passing the other dependencies like price_final_daily directly to the method with dataframe from memory.  But I wanted to debug and inspect the context variables values in the debugger to understand how weekly partitionned assets get their data from daily assets in dagster.

**Reply by sryza** [reply]
Here's a minimal example that I found reproduces this: 

```python
from dagster import asset, SourceAsset, materialize_to_memory


@asset
def price_weekly(price_final_daily):
    ...


if __name__ == "__main__":
    result = materialize_to_memory([SourceAsset("price_final_daily"), price_weekly])
```

**Comment by sryza** [comment, marked as answer]
The error message is confusing, but here's the issue: Dagster doesn't know how to load the value of the source asset, and there currently isn't a great way to tell it to do so with `materialize_to_memory`. Here's a Github issue where we're tracking making this easier: https://github.com/dagster-io/dagster/issues/7656.

You might consider creating mock asset definitions that return the values that you want to use in your test:

```python
from dagster import asset, materialize_to_memory

@asset
def upstream():
    ...


@asset
def downstream(upstream):
    ...

@asset(name="upstream")
def mock_upstream():
    return some_hardcoded_dataframe_for_use_in_my_test

materialize_to_memory([downstream, mock_upstream])
```

**Reply by EtienneT** [reply]
Thanks for the clarification!  The mock makes sense.

Because I tried your example here and got the same error: https://github.com/dagster-io/dagster/discussions/14630#discussioncomment-6115680

My main use case for this is not really to test my code, but since I am learning dagster, I wanted to debug things and inspect the variables in the debugger.  I specifically wanted to test different values of partition mapping with my things and see what value I got back etc.

What is the recommended way of debugging pipelines in dagster?

Thanks,

**Reply by sryza** [reply]
Have you considered using `materialize`? The values that you get back with partition mappings can depend on your IO manager, and `materialize` will invoke your IO manager, unlike `materialize_to_memory`.
